### ISSUE TITLE
Vagrant fails with "libvirt is not a recognized provider"

### ISSUE THREAD

**Issue by dolobanko** [vagrant-libvirt/vagrant-libvirt#770]
I do have strange behavior with vagrant-libvirt plugin on Centos 6.8, libvirt v0.10.2.

### Steps to reproduce
1. Install the latest Vagrant binary (1.9.4).
2. Install vagrant-libvirt plugin: `vagrant plugin install vagrant-libvirt`. The strange thing here, that plugin is version 0.0.20.
3. Initialize vagrant file: `vagrant init`.
4. Run `vagrant up --provisioner=libvirt`

### Expected behaviour
VM up and running
### Actual behaviour
Below stacktrace:

**vagrant up --provider=libvirt**  

/root/.vagrant.d/gems/2.2.5/gems/fog-core-1.43.0/lib/fog/core/services_mixin.rb:12:in `new': libvirt is not a recognized provider (ArgumentError)
        from /root/.vagrant.d/gems/2.2.5/gems/fog-core-1.43.0/lib/fog/compute.rb:54:in `new'
        from /root/.vagrant.d/gems/2.2.5/gems/vagrant-libvirt-0.0.20/lib/vagrant-libvirt/action/connect_libvirt.rb:38:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/action/warden.rb:34:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/action/builtin/config_validate.rb:25:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/action/warden.rb:34:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/action/builder.rb:116:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/action/runner.rb:66:in `block in run'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/util/busy.rb:19:in `busy'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/action/runner.rb:66:in `run'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/machine.rb:225:in `action_raw'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/machine.rb:200:in `block in action'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/environment.rb:567:in `lock'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/machine.rb:186:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/machine.rb:186:in `action'
        from /root/.vagrant.d/gems/2.2.5/gems/vagrant-libvirt-0.0.20/lib/vagrant-libvirt/provider.rb:58:in `state'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/machine.rb:506:in `state'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/machine.rb:145:in `initialize'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/vagrantfile.rb:79:in `new'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/vagrantfile.rb:79:in `machine'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/environment.rb:669:in `machine'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/plugin/v2/command.rb:177:in `block in with_target_vms'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/plugin/v2/command.rb:201:in `call'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/plugin/v2/command.rb:201:in `block in with_target_vms'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/plugin/v2/command.rb:183:in `each'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/plugin/v2/command.rb:183:in `with_target_vms'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/plugins/commands/up/command.rb:131:in `install_providers'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/plugins/commands/up/command.rb:85:in `execute'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/cli.rb:42:in `execute'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/lib/vagrant/environment.rb:308:in `cli'
        from /opt/vagrant/embedded/gems/gems/vagrant-1.9.4/bin/vagrant:127:in `<main>'

### System configuration
**OS/Distro version:**: Centos 6.8

**Libvirt version:** v0.10.2 

**Output of `vagrant version; vagrant plugin list`:**

`vagrant version
Installed Version: 1.9.4
Latest Version: 1.9.4

You're running an up-to-date version of Vagrant!

 vagrant plugin list
vagrant-libvirt (0.0.20)
  - Version Constraint: > 0
vagrant-share (1.1.7)
  - Version Constraint: > 0
`

**Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt`**
The same as above stacktrace.
``

**A Vagrantfile to reproduce the issue:**
```
Vagrant.configure("2") do |config|
  config.vm.box = "centos/7"
end
```

**Comment by timhughes** [comment]
seeing the same thing

**Comment by timhughes** [comment]
#769 will fix this issue 

for the time being this works
```
mkdir -p ~/tmp/
cd ~/tmp/
git clone https://github.com/jcf/vagrant-libvirt
cd ./vagrant-libvirt/
git checkout upgrade-nokogiri
gem build vagrant-libvirt.gemspec
vagrant plugin install  ~/tmp/vagrant-libvirt/vagrant-libvirt-0.0.37.gem
```

**Comment by andrewpgit** [comment]
Thank you. It solved the issue.

**Comment by SturmDrangAltaar** [comment]
Just to add to the above workaround. I didn't have ruby installed natively on my system, so I used the gem binary located at `/opt/vagrant/embedded/bin/gem`

I don't know if there are any downsides to doing this, but I didn't want to install and configure ruby to get this right. Gem file that was generated was named differently as `vagrant-libvirt-0.0.40.gem` but worked all the same and my vagrants are running again.

Apologies if anything I've mentioned here is incorrect, my experience with ruby is very limited.

**Comment by alexggolovin** [comment]
Issue is fixed on Debian 8.8 with vagrant-libvirt (0.0.40) installed in the way proposed by timhughes, but new dhcp leases issue is implementing https://github.com/vagrant-libvirt/vagrant-libvirt/issues/674 and temporary fixed with the workaround proposed by mrvovanness here https://github.com/fog/fog-libvirt/issues/28

**Comment by infernix** [comment]
Can you please retest with 1.9.7?

**Comment by davidcomeyne** [comment]
I have the same issue with Vagrant 1.9.7.
Tried the fix proposed by @timhughes, to no avail (it used vagrant-libvirt-0.0.40.gem).

This is the output when installing the .gem:
`Installing the 'vagrant-libvirt-0.0.40.gem' plugin. This can take a few minutes...
Fetching: nokogiri-1.7.2.gem (100%)
Building native extensions.  This could take a while...
Building native extensions.  This could take a while...
Bundler, the underlying system Vagrant uses to install plugins,
reported an error. The error is shown below. These errors are usually
caused by misconfigured plugin installations or transient network
issues. The error from Bundler is:

ERROR: Failed to build gem native extension.

    current directory: /home/david/.vagrant.d/gems/2.3.4/gems/ruby-libvirt-0.7.0/ext/libvirt
/opt/vagrant/embedded/bin/ruby -r ./siteconf20170818-12534-kxruud.rb extconf.rb
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/vagrant/embedded/bin/$(RUBY_BASE_NAME)
	--with-libvirt-include
	--without-libvirt-include
	--with-libvirt-lib
	--without-libvirt-lib
	--with-libvirt-config
	--without-libvirt-config
	--with-pkg-config
	--without-pkg-config
extconf.rb:73:in `<main>': libvirt library not found in default locations (RuntimeError)

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /home/david/.vagrant.d/gems/2.3.4/extensions/x86_64-linux/2.3.0/ruby-libvirt-0.7.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /home/david/.vagrant.d/gems/2.3.4/gems/ruby-libvirt-0.7.0 for inspection.
Results logged to /home/david/.vagrant.d/gems/2.3.4/extensions/x86_64-linux/2.3.0/ruby-libvirt-0.7.0/gem_make.out`

The mkmf.log states:
`"pkg-config --exists libvirt"
package configuration for libvirt is not found`

Could easily be an issue on my system (Fedora 26), but if you do know a possible solution I would love to try it out.

**Comment by HartS** [comment]
Looking at your 4th step, where are you getting the `--provisioner` flag? I don't see it documented by vagrant anywhere.. I've been using `--provider=libvirt`

**Comment by electrofelix** [comment]
@HartS reasonably sure it's a type and should have been `--provider=libvirt`, `--provision` and `--provision-with=shell` would be the correct arguments if you were looking to control that. The command run and output is consistent with the option having been '--provider=libvirt`.

**Comment by pag-r** [comment]
I had the same issue on my Fedora 26. No solutions was working except one, taken from ArchLinux Vagant wiki: [https://wiki.archlinux.org/index.php/Vagrant#vagrant-libvirt](https://wiki.archlinux.org/index.php/Vagrant#vagrant-libvirt). Package libvirt-devel is required. Vagrant package: vagrant-2.0.0-1.x86_64

**Comment by friendlypenguin** [comment]
libvirt-dev for debian/ubuntu users

**Comment by dolorsitatem** [comment]
Running Debian 9 (Stretch) and installing `libvirt-dev` solved it for me.

**Comment by timhughes** [comment]
This was fixed in #769  so this ticket can be closed
